extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/CommonCacheNotifier.java

```diff
@@ -0,0 +1,238 @@
+  )
+  {
+    this.threadName = threadName;
+    this.logger = logger;
+    this.itemsToUpdate = new HashSet<>();
+    this.itemConfigMap = itemConfigMap;
+    this.serializedMaps = new HashMap<>();
+    this.discoveryProvider = discoveryProvider;
+    this.httpClient = httpClient;
+    this.baseUrl = baseUrl;
+  }
+
+  public void start()
+  {
+    notifierThread = Execs.makeThread(
+        threadName,
+        () -> {
+          while (!Thread.interrupted()) {
+            try {
+              logger.debug("Waiting for cache update notification");
+              Set<String> itemsToUpdateSnapshot;
+              HashMap<String, byte[]> serializedMapsSnapshot;
+              synchronized (itemsToUpdate) {
+                if (itemsToUpdate.isEmpty()) {
+                  itemsToUpdate.wait();
+                }
+                itemsToUpdateSnapshot = new HashSet<>(itemsToUpdate);
+                serializedMapsSnapshot = new HashMap<String, byte[]>(serializedMaps);
+                itemsToUpdate.clear();
+                serializedMaps.clear();
+              }
+              logger.debug("Sending cache update notifications");
+              for (String authorizer : itemsToUpdateSnapshot) {
+                BasicAuthDBConfig authorizerConfig = itemConfigMap.get(authorizer);
+                if (!authorizerConfig.isEnableCacheNotifications()) {
+                  continue;
+                }
+
+                // Best effort, if a notification fails, the remote node will eventually poll to update its state
+                // We wait for responses however, to avoid flooding remote nodes with notifications.
+                List<ListenableFuture<StatusResponseHolder>> futures = sendUpdate(
+                    authorizer,
+                    serializedMapsSnapshot.get(authorizer)
+                );
+                for (ListenableFuture<StatusResponseHolder> future : futures) {
+                  try {
+                    StatusResponseHolder srh = future.get(
+                        authorizerConfig.getCacheNotificationTimeout(), TimeUnit.MILLISECONDS
+                    );
+                    logger.debug("Got status: " + srh.getStatus());
+                  }
+                  catch (Exception e) {
+                    logger.makeAlert(e, "Failed to get response for cache notification.").emit();
+                  }
+                }
+              }
+              logger.debug("Received responses for cache update notifications.");
+            }
+            catch (Throwable t) {
+              logger.makeAlert(t, "Error occured while handling updates for cachedUserMaps.").emit();
+            }
+          }
+        },
+        true
+    );
+    notifierThread.start();
+  }
+
+  public void addUpdate(String updatedItemName, byte[] updatedItemData)
+  {
+    synchronized (itemsToUpdate) {
+      itemsToUpdate.add(updatedItemName);
+      serializedMaps.put(updatedItemName, updatedItemData);
+      itemsToUpdate.notify();
```

Comment:
I think the current behavior is okay for now, since the updates should be pretty infrequent, I changed this to use Pair<String, byte[]> and a BlockingQueue